### ISSUE TITLE
DMSDK job fails with mismatching host names

### ISSUE THREAD

**Issue by sastafford** [marklogic/data-movement#143]
TL/DR - I'm passing in a host name into the DatabaseClient object, it's being ignored and the host name in the MarkLogic Admin console is being used instead.  

I am running a MarkLogic docker container where the hostname is a given alphanumeric string.  i.e. 284472306b64

If I look at the hosts in the ML admin console the host name is "284472306b64.localdomain"

Now, I'm running a some unit tests JVM on the host machine and not the docker container so I'm setting the DatabaseClient hostname to "localhost".   I've also hardcoded the IP address of the machine.  The DMSDK job that I've created throws a UnknownHostException for some reason.  

 com.marklogic.datamovement.QueryHostException: com.sun.jersey.api.client.ClientHandlerException: java.net.UnknownHostException: grover: Name or service not known
        at com.marklogic.datamovement.impl.QueryHostBatcherImpl$QueryTask.run(QueryHostBatcherImpl.java:299)
        at com.marklogic.datamovement.impl.QueryHostBatcherImpl.startQuerying(QueryHostBatcherImpl.java:217)
        at com.marklogic.datamovement.impl.QueryHostBatcherImpl.start(QueryHostBatcherImpl.java:174)
        at com.marklogic.datamovement.impl.DataMovementServices.startJob(DataMovementServices.java:71)
        at com.marklogic.datamovement.DataMovementManager.startJob(DataMovementManager.java:97)

**Comment by sastafford** [comment]
Other reference: https://circleci.com/gh/sastafford/marklogic-spring-batch/30

**Comment by sammefford** [comment]
Can you give me more information?  Are you using more than one node?  What version of the server are you using?  We have a fix in the latest server that uses your provided hostname for that host.  But if you're talking to a cluster you'll have to make sure the java client host can resolve the hostnames of each server in the cluster.  This is just like MLCP.

**Comment by sammefford** [comment]
Could you please try it with EA4?  It's available here: http://developer.marklogic.com/maven2/com/marklogic/java-client-api/

**Comment by sastafford** [comment]
I tried with Java Client EA4 and DMSDK EA3 and I got the following error.  I did not see a DMSDK EA4 out on the ML maven site. 

java.lang.NoSuchMethodError: com.marklogic.client.DatabaseClient.getUser()Ljava/lang/String;
	at com.marklogic.datamovement.impl.ForestConfigurationImpl.getForestClient(ForestConfigurationImpl.java:59)
	at com.marklogic.datamovement.impl.QueryHostBatcherImpl.initialize(QueryHostBatcherImpl.java:191)
	at com.marklogic.datamovement.impl.QueryHostBatcherImpl.start(QueryHostBatcherImpl.java:172)
	at com.marklogic.datamovement.impl.DataMovementServices.startJob(DataMovementServices.java:71)
	at com.marklogic.datamovement.DataMovementManager.startJob(DataMovementManager.java:97)
	at com.marklogic.spring.batch.item.tasklet.DeleteDocumentsTasklet.execute(DeleteDocumentsTasklet.java:37)

**Comment by sammefford** [comment]
Please just use Java Client EA4.  As of EA4 we've merged DMSDK into Java Client.

**Comment by jmakeig** [comment]
The EA builds are in our own Maven repo. GA builds will be on Maven Central and JCenter.

```groovy
dependencies {
  repositories {
    mavenCentral()
    mavenLocal()
    jcenter()
    maven { 
      url 'https://developer.marklogic.com/maven2'
    }
  }
  compile 'com.marklogic:marklogic-client-api:4.0-EA4'
  // …
}
```

**Comment by jmakeig** [comment]
Usually for pre-releases, I just compile and install the `develop` branch of the Java Client in my local Maven repo and use the `SNAPSHOT` artifact id. That’s more of a Java Client API developer set-up, though.

**Comment by sastafford** [comment]
Updated my jar file to EA4 and modified my DMSDK code given the package/class renaming.  Here is my [code](https://github.com/sastafford/marklogic-spring-batch/blob/feature/190-upgradeJavaClient/infrastructure/src/main/java/com/marklogic/spring/batch/item/tasklet/DeleteDocumentsTasklet.java) that deletes a bunch of documents.

I'm still getting a host exception with the code change.  My host name for the DatabaseClient (localhost) differs from the host name being assigned to my MarkLogic host (some alphanumeric string assigned by my docker container).  When I modify the MarkLogic host to "localhost", the test case executes correctly.  

So it comes down to host name differences between the DatabaseClient and the MarkLogic Admin.

**Comment by jmakeig** [comment]
@sammefford, isn't this a use case for `FilteredForestConfiguration`?

**Comment by jmakeig** [comment]
Actually, I take that back. This sounds like some link in the chain not understanding how to correctly resolve the hostname. This is the default behavior on macOS when you don’t rename the host to `localhost`. I wouldn’t imagine that the Java Client is doing something unorthodox to resolve names, though.

**Comment by kcoleman-marklogic** [comment]
Or the DMSDK variant of this problem?

http://docs.marklogic.com/guide/mlcp/troubleshoot#id_27997

**Comment by jmakeig** [comment]
> Or the DMSDK variant of this problem?

Yes, probably. I don’t know how to tell my networking stack (OS + Java) how to correctly resolve `localhost`, though. This seems pretty fundamental, no?

**Comment by kcoleman-marklogic** [comment]
If this is similar to issues mlcp users run into, @jxchen-us could speak to this much more authoritatively, but I believe you have to use the hostname that is resolvable, which might or might not be localhost. You might need to change what hostname ML is configured as if it's localhost and that's not resolvable.

Similarly, mlcp fetches forest host names from ML so it can connect directly to them. The names returned by ML must be resolvable on the client. IDK if that applies to DMSDK or not.

**Comment by jmakeig** [comment]
Yes, I understand conceptually, but when I fire up the Java Client or mlcp on my laptop and it doesn’t connect to `localhost` without extra intervention I get bemused.

**Comment by jxchen-us** [comment]
mlcp maps your original user-specified host name with the MarkLogic-configured host name as long as your MarkLogic cluster contains single node.  Otherwise, you need to use -restrict_host.

**Comment by sastafford** [comment]
This should be pretty easy to create.  Pass in "localhost" to your DatabaseClient and then rename your MarkLogic host a different name.  See if it works.

**Comment by jmakeig** [comment]
> This should be pretty easy to create. Pass in "localhost" to your DatabaseClient and then rename your MarkLogic host a different name. See if it works.

Which is the default scenario on a virgin install of MarkLogic, Java, and mlcp/Java Client API. (At least on my macOS machine.) My browser, for example, doesn’t need extra intervention to know how to find `localhost`. This seems to be a Java thing.

**Comment by sastafford** [comment]
I want to remind everyone on this thread, that I'm running with a docker container for this test.  Thanks for the comments so far.

**Comment by sammefford** [comment]
I'm sorry, Scott, I should have also asked you to use the EA4 server release.  That's actually where the renaming happens, in the /v1/internal/forestinfo endpoint.  It takes the hostname you used to call it, and provides that back to you as the hostname to use for that node in the cluster.  Like MLCP, this solution only works for a single-node cluster.  For a multi-node cluster, you'll have to rename your hosts to addressable names or use FilteredForestConfig.

**Comment by jmakeig** [comment]
With MarkLogic 9.0-EA4(is, 20170120), marklogic-client-api-4.0.0-EA4, Java 1.8.0_111-b14, macOS 10.12.2 (16C67), I’m able to connect to `localhost` without renaming my MarkLogic host. (I’m pretty sure this didn’t work at sometime in the past.) Thus, I’ll shutdown this distraction so we can focus on @sastafford’s actual issue. Sorry for the noise.

**Comment by sastafford** [comment]
I've confirmed that with MarkLogic 9.0-EA4, java-client-api:4.0.0-EA4 this issue is resolved.
